_posts/2021-01-12-feedback-to-users-requests-in-distributed-systems.md

```diff
@@ -0,0 +1,105 @@
+
+> Q: To which webserver instance?
+
+If web servers are scaled out and use a queue to communicate with backend services, there is no guarantee that the backend service response message goes to the same web server instance that sent out the initial request. If there was no user, and no WebSockets, involved we had no problems, scaled out instances can be designed to be stateless, and thus, what instance handles a message is not that important. A device using WebSocket connects directly to a specific web server instance. In such a scenario, if the user sends the initial request to web server instance A, and web server instance B is handling the response message from the backend service, there is no user's WebSocket connected to instance B. The status is lost because web server instance B is missing some context or information to talk back to the user's device.
+
+## Backplanes to the rescue.
+
+The solution to this second set of issues is to use a backplane. Backplanes are a side communication channel a set of scaled out service instances can use to communicate among themselves. From the main communication channel perspective, the one used to share information across the entire system, a set of scaled out instances wants to appear as a single logical unit. For this reason, they are usually referred to as a logical service. From a sender's perspective, it doesn't really matter which instance will handle a request; they route the request to the logical destination. The same happens from a recipient's perspective. It doesn't matter which instance sent the request; most of the time, it's the logical sender that is the only thing that matters. There are cases, such as the WebSockets one, in which the physical instances matter a lot. A failure in addressing the correct instance breaks the whole communication chain.
+
+A simple solution would be to address the response to the webserver instance that initially handled the user request. However, that's not so simple. The web server instance might not exist anymore; the user WebSocket connection might have dropped and reconnected to a different web server instance.
+Many things can happen to make routing the response to a specific instance less effective than expected. Not to say that allowing senders to route to a specific instance brings to the table one of the distributed systems fallacies: topology doesn't change. Topology changes all the time, and there is no easy and cheap way to guarantee that recipients will be where senders expect them to be.
+
+Backplanes allow hiding away from senders the recipients' topology. When one of the scaled-out instances receives a message from the system, it notifies all the other instances through the backplane. The backplane could be using the exact same mechanism the system uses to communicate to peers in the logical service; it could be using a message/queue-based approach or a completely different one, for example, Redis pub/sub.
```

Comment:
```suggestion
Backplanes hide the recipients' topology from senders. When one of the scaled-out instances receives a message from the system, it notifies all the other instances through the backplane. The backplane could be using the exact same mechanism the system uses to communicate to peers in the logical service; it could be using a message/queue-based approach or a completely different one, for example, Redis pub/sub.
```